consensus/hotstuff/validator/validator.go

```diff
@@ -12,74 +12,140 @@ import (
+// ValidateTC validates the TimeoutCertificate `tc`.
 // During normal operations, the following error returns are expected:
-//  * model.InvalidBlockError if the QC is invalid
-func (v *Validator) ValidateQC(qc *flow.QuorumCertificate, block *model.Block) error {
-	if qc.BlockID != block.BlockID {
-		// Sanity check! Failing indicates a bug in the higher-level logic
-		return fmt.Errorf("qc.BlockID %s doesn't match block's ID %s", qc.BlockID, block.BlockID)
+//  * model.InvalidTCError if the TC is invalid
+//  * model.ErrViewForUnknownEpoch if the TC refers unknown epoch
+// Any other error should be treated as exception
+func (v *Validator) ValidateTC(tc *flow.TimeoutCertificate) error {
+	highestQC := tc.NewestQC
+
+	// The TC's view cannot be smaller than the view of the QC it contains.
+	// Note: we specifically allow for the TC to have the same view as the highest QC.
+	// This is useful as a fallback, because it allows replicas other than the designated
+	// leader to also collect votes and generate a QC.
+	if tc.View < highestQC.View {
+		return newInvalidTCError(tc, fmt.Errorf("TC's QC cannot be newer than the TC's view"))
+	}
+
+	// verifying that tc.NewestQC is the QC with the highest view
+	highestQCView := tc.NewestQCViews[0]
+	for _, view := range tc.NewestQCViews {
+		if highestQCView < view {
+			highestQCView = view
+		}
+	}
+	if highestQCView != tc.NewestQC.View {
+		return newInvalidTCError(tc, fmt.Errorf("included QC (view=%d) should be equal to highest contributed view: %d", tc.NewestQC.View, highestQCView))
+	}
+
+	// 1. Check if there is super-majority of votes
+	allParticipants, err := v.committee.IdentitiesByEpoch(tc.View)
+	if err != nil {
+		return fmt.Errorf("could not get consensus participants at view %d: %w", tc.View, err)
+	}
+	signers, err := signature.DecodeSignerIndicesToIdentities(allParticipants, tc.SignerIndices)
+	if err != nil {
+		if signature.IsDecodeSignerIndicesError(err) {
+			return newInvalidTCError(tc, fmt.Errorf("invalid signer indices: %w", err))
+		}
+		// unexpected error
+		return fmt.Errorf("unexpected internal error decoding signer indices: %w", err)
+	}
+
+	// determine whether signers reach minimally required weight threshold for consensus
+	threshold, err := v.committee.WeightThresholdForView(tc.View)
+	if err != nil {
+		return fmt.Errorf("could not get weight threshold for view %d: %w", tc.View, err)
+	}
+	if signers.TotalWeight() < threshold {
+		return newInvalidTCError(tc, fmt.Errorf("tc signers have insufficient weight of %d (required=%d)", signers.TotalWeight(), threshold))
 	}
-	if qc.View != block.View { // check view
-		return newInvalidBlockError(block, fmt.Errorf("qc's View %d doesn't match referenced block's View %d", qc.View, block.View))
+
+	// Validate QC
+	err = v.ValidateQC(highestQC)
+	if err != nil {
+		if model.IsInvalidQCError(err) {
+			return newInvalidTCError(tc, fmt.Errorf("invalid QC included in TC: %w", err))
+		}
```

Comment:
I think we also require special handling of `ErrViewForUnknownEpoch` here:

```suggestion
		}
				if errors.Is(err, model.ErrViewForUnknownEpoch) {
			// We require each replica to be bootstrapped with a QC pointing to a finalized block. Consensus safety rules guarantee that
			// a QC at least as new as the root QC must be contained in any TC. This is because the TC must include signatures from a
			// supermajority of replicas, including at least one honest replica, which attest to their locally highest known QC. Hence, 
			// any QC included in a TC must be the root QC or newer. Therefore, we should know the Epoch for any QC we encounter.
			// receiving a `model.ErrViewForUnknownEpoch` is conceptually impossible, i.e. a symptom of an internal bug or invalid
			// bootstrapping information. 
			return fmt.Errorf("no Epoch information availalbe for QC that was included in TC; symptom of internal bug or invalid bootstrapping information: %s", err.Error())
		}
```